src/v/storage/log_manager.cc

```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
      * them as fast as we can since the majority of the time they will be in the
      * hotpath / (request-response)
      */
+    std::unique_ptr<model::ntp> current_key{};
     using bflags = log_housekeeping_meta::bitflags;
-    return ss::do_until(
-             [this] {
-                 auto it = find_next_non_compacted_log(_logs);
-                 return it == _logs.end();
-             },
-             [this, collection_threshold] {
-                 return ss::with_scheduling_group(
-                   _config.compaction_sg, [this, collection_threshold] {
-                       auto it = find_next_non_compacted_log(_logs);
-                       if (it == _logs.end()) {
-                           // must check again because between the stop
-                           // condition and this continuation we might have
-                           // removed the log
-                           return ss::now();
-                       }
-                       it->second.flags |= bflags::compacted;
-                       it->second.last_compaction = ss::lowres_clock::now();
-                       return it->second.handle.compact(compaction_config(
+    return ss::repeat([this,
+                       collection_threshold,
+                       current_key = std::move(current_key)]() mutable {
```

Comment:
it is generally best practice to use `ss::do_with` to hold state that is going to live across loop iterations. for example

`ss::do_with(Type t, [](Type& instance) { return ss::repeat([&instance, ....`

although if you switch to coroutines then this state can live on the stack.